ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -1,397 +1,385 @@
-        (blockFetchDecisionTracer tracers)
-        (blockFetchClientTracer   tracers)
-        blockFetchInterface
-        fetchClientRegistry
-
-    return NodeKernel
-      { getChainDB             = chainDB
-      , getMempool             = mempool
-      , getNodeConfig          = cfg
-      , getFetchClientRegistry = fetchClientRegistry
-      , getNodeCandidates      = varCandidates
-      , getTracers             = tracers
+import qualified Ouroboros.Consensus.Util.ResourceRegistry as ResourceRegistry
+
+import           Ouroboros.Storage.ChainDB (ChainDB, ChainDbArgs)
+import qualified Ouroboros.Storage.ChainDB as ChainDB
+import           Ouroboros.Storage.EpochInfo (newEpochInfo)
+import           Ouroboros.Storage.ImmutableDB (ValidationPolicy (..))
+import           Ouroboros.Storage.LedgerDB.DiskPolicy (defaultDiskPolicy)
+import           Ouroboros.Storage.LedgerDB.MemPolicy (defaultMemPolicy)
+
+-- | Start a node.
+--
+-- This opens the 'ChainDB', sets up the 'NodeKernel' and initialises the
+-- network layer.
+--
+-- This function runs forever unless an exception is thrown.
+run
+  :: forall blk peer. (RunNode blk, Ord peer)
+  => Tracers IO peer blk                  -- ^ Consensus tracers
+  -> Tracer  IO (ChainDB.TraceEvent blk)  -- ^ ChainDB tracer
+  -> RunNetworkArgs peer blk              -- ^ Network args
+  -> FilePath                             -- ^ Database path
+  -> ProtocolInfo blk
+  -> (ChainDbArgs IO blk -> ChainDbArgs IO blk)
+      -- ^ Customise the 'ChainDbArgs'
+  -> (NodeArgs IO peer blk -> NodeArgs IO peer blk)
+      -- ^ Customise the 'NodeArgs'
+  -> (ResourceRegistry IO -> NodeKernel IO peer blk -> IO ())
+     -- ^ Called on the 'NodeKernel' after creating it, but before the network
+     -- layer is initialised.
+  -> IO ()
+run tracers chainDbTracer rna dbPath pInfo
+    customiseChainDbArgs customiseNodeArgs onNodeKernel =
+    ResourceRegistry.with $ \registry -> do
+
+      chainDB <- initChainDB
+        chainDbTracer
+        registry
+        dbPath
+        cfg
+        initLedger
+        slotLength
+        customiseChainDbArgs
+
+      btime <- realBlockchainTime
+        registry
+        slotLength
+        (nodeStartTime (Proxy @blk) cfg)
+
+      let nodeArgs = customiseNodeArgs $ mkNodeArgs
+            registry
+            cfg
+            initState
+            tracers
+            btime
+            chainDB
+
+      nodeKernel <- initNodeKernel nodeArgs
+
+      onNodeKernel registry nodeKernel
+
+      initNetwork
+        registry
+        nodeArgs
+        nodeKernel
+        rna
+  where
+    ProtocolInfo
+      { pInfoConfig     = cfg
+      , pInfoInitLedger = initLedger
+      , pInfoInitState  = initState
+      } = pInfo
+
+    -- TODO this will depend on the protocol and can change when a hard-fork
+    -- happens, see #921 and #282.
+    --
+    -- For now, we hard-code it to Byron's 20 seconds.
+    slotLength = slotLengthFromMillisec (20 * 1000)
```

Comment:
This still needs to be addressed.

Reply:
We've concluded we will do this in a follow-up PR.